generator/sidekick/downloads_cache_test.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/walle/targz"
+)
+
+func TestExistingDirectory(t *testing.T) {
+	tmp, err := os.MkdirTemp(t.TempDir(), "sidekick-test-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmp)
+	rootConfig := Config{
+		Source: map[string]string{
+			"googleapis-root": tmp,
+		},
+	}
+	root, err := makeGoogleapisRoot(&rootConfig)
+	if err != nil {
+		t.Error(err)
+	}
+	if root != tmp {
+		t.Errorf("mismatched root directory got=%s, want=%s", root, tmp)
+	}
+}
+
+func TestValidateConfig(t *testing.T) {
+	rootConfig := Config{
+		Source: map[string]string{
+			"googleapis-root": "https://unused",
+		},
+	}
+	_, err := makeGoogleapisRoot(&rootConfig)
+	if err == nil {
+		t.Errorf("expected error when missing `googleapis-sha256")
+	}
+}
+
+func TestWithDownload(t *testing.T) {
+	testDir, err := os.MkdirTemp(t.TempDir(), "sidekick-test-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(testDir)
+
+	simulatedSha := "2d08f07eab9bbe8300cd20b871d0811bbb693fab"
+	simulatedSubdir := fmt.Sprintf("googleapis-%s", simulatedSha)
+	simulatedPath := fmt.Sprintf("/archive/%s.tar.gz", simulatedSha)
+	tarball, err := makeTestTarball(t, testDir, simulatedSubdir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// In this test we expect that a download is needed.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != simulatedPath {
+			t.Errorf("Expected to request '%s', got: %s", simulatedPath, r.URL.Path)
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write(tarball.Contents)
+	}))
+	defer server.Close()
+
+	rootConfig := &Config{
```

Comment:
Just for my understanding, how is this rootConfig created? How is the sha256 available ahead?

Reply:
Eventually we will have a `.sidekick.toml` file in the top-level directory of the monorepo. The URL and SHA256 is stored in that file.  At first we will update those manually, eventually `sidekick upgrade` will update both.

What other magic happens on those upgrades is up for debate.